src/danfojs-base/core/frame.ts

```diff
@@ -1333,6 +1333,106 @@ export default class DataFrame extends NDframe implements DataFrameInterface {
+     * console.log(df1)
+     * 
+     * // Difference with previous 3rd previous row
+     * const df2 = df.diff(3)
+     * console.log(df2)
+     * 
+     * // Difference with following row
+     * const df3 = df.diff(-1)
+     * console.log(df3)
+     * 
+     * // Difference with another DataFrame
+     * const df4 = df.diff(df3)
+     * console.log(df4)
+     * ```
+    */
+    diff(other: DataFrame | Series | number[] | number, options?: { axis?: 0 | 1, inplace?: boolean }): DataFrame
+    diff(other: DataFrame | Series | number[] | number, options?: { axis?: 0 | 1, inplace?: boolean }): DataFrame | void {
+        const { inplace, axis } = { inplace: false, axis: 1, ...options }
+
+        if (this.$frameIsNotCompactibleForArithmeticOperation()) {
+            throw Error("TypeError: diff operation is not supported for string dtypes");
+        }
+
+        if ([0, 1].indexOf(axis) === -1) {
+            throw Error("ParamError: Axis must be 0 or 1");
+        }
+
+        if (other === 0) {
+            return this;
+        }
+
+        if (typeof other === "number" && axis === 1) {
+            const orig_tensor = this.tensor.clone();
+            let unit = [NaN];
+            for (let i = 1; i < orig_tensor.shape[orig_tensor.rank - 1]; i++) {
+                unit.push(NaN);
+            }
+            let diff_array: any[] = orig_tensor.arraySync();
+            if (other > 0) {
+                for (let i = 0; i < other; i++) {
+                    diff_array.unshift(unit);
+                    diff_array.pop();
+                }
+            }
+            else if (other < 0) {
+                for (let i = 0; i > other; i--) {
+                    diff_array.push(unit);
+                    diff_array.shift();
+                }
+            }
+            const diff_tensor = tensorflow.tensor2d(diff_array, orig_tensor.shape);
+            return this.$MathOps([orig_tensor, diff_tensor], "sub", inplace);
+        }
+
+        if (typeof other === "number" && axis === 0) {
+            const orig_df = new DataFrame(this.tensor.clone());
+            const orig_tensor = orig_df.T.tensor.clone();
+            let unit = [NaN];
+            for (let i = 1; i < orig_tensor.shape[orig_tensor.rank - 1]; i++) {
+                unit.push(NaN);
+            }
+            let diff_array: any[] = orig_tensor.arraySync();
+            if (other > 0) {
+                for (let i = 0; i < other; i++) {
+                    diff_array.unshift(unit);
+                    diff_array.pop();
+                }
+            }
+            else if (other < 0) {
+                for (let i = 0; i > other; i--) {
+                    diff_array.push(unit);
+                    diff_array.shift();
+                }
+            }
+            const diff_tensor = tensorflow.tensor2d(diff_array, orig_tensor.shape);
+            const diff_df = this.$MathOps([orig_tensor, diff_tensor], "sub", inplace) as DataFrame;
+            return diff_df.T;
```

Comment:
Suggestion: If this is the only difference between when axis=1 or 0, then a cleaner method will be to perform the computation and at the end based on the axis you can return the transpose:

```javascript
// perform all computation ...
// ...

if (axis == 0) {
  return diff_df
} else {
  return diff_df.T;
}



```

Reply:
The problem with doing this is that I'm essentially flipping the entire thing with `df.T` first and then pretending the columns are rows, then flipping them back at the end.

I'll try to refactor it so it's more succinct and less duplicated!